File: server/src/routes/items.ts
import { Hono } from 'hono';
import { authMiddleware } from '../auth';
import { r2ConfigForWorld, r2DeleteObject, itemKeyToObjectKey } from '../r2';
import { versionCompare } from '../version';
import { db } from '../db';
import { log } from '../logger';
import type { AuthContext } from '../auth';

type Variables = { auth: AuthContext };

export const itemsRouter = new Hono<{ Variables: Variables }>();

// POST /items/query — latest version for a set of keys (exact or * wildcard)
itemsRouter.post('/query', authMiddleware(1), async (c) => {
  const auth = c.get('auth');
  const body = await c.req.json<{ keys: string[] }>();
  const { keys } = body;

  if (!Array.isArray(keys)) {
    return c.json({ items: [] });
  }
  if (keys.length > 1000) {
    return c.json({ error: 'max 1000 keys per request' }, 400);
  }

  const filtered = keys.filter(k => k !== '');
  const exactKeys = filtered.filter(k => !k.includes('*'));
  const likePatterns = filtered.filter(k => k.includes('*')).map(k => k.replace(/\*/g, '%'));

  type ItemRow = { key: string; version: string; created_at: number };

  let sql = `SELECT key, version, created_at FROM items WHERE world = ?`;
  const bindings: unknown[] = [auth.world];

  if (filtered.length > 0) {
    const conditions: string[] = [];
    if (exactKeys.length > 0) {
      conditions.push(`key IN (${exactKeys.map(() => '?').join(', ')})`);
      bindings.push(...exactKeys);
    }
    for (const pattern of likePatterns) {
      conditions.push('key LIKE ?');
      bindings.push(pattern);
    }
    sql += ` AND (${conditions.join(' OR ')})`;
  }

  log.debug(`items/query: world=${auth.world} keys=${JSON.stringify(filtered)}`);

  const rows = db.prepare(sql).all(...bindings) as ItemRow[];

  const byKey = new Map<string, ItemRow>();
  for (const row of rows) {
    const current = byKey.get(row.key);
    if (!current || versionCompare(row.version, current.version) > 0) {
      byKey.set(row.key, row);
    }
  }

  const items = [...byKey.values()].sort((a, b) => a.key.localeCompare(b.key));
  return c.json({ items });
});

// GET /items/:key — all versions for a key, newest first
itemsRouter.get('/:key', authMiddleware(1), async (c) => {
  const auth = c.get('auth');
  const key = decodeURIComponent(c.req.param('key'));

  log.debug(`items/get: world=${auth.world} key=${key}`);

  const results = db.prepare(
    `SELECT version, created_at FROM items WHERE world = ? AND key = ?`
  ).all(auth.world, key) as { version: string; created_at: number }[];

  if (results.length === 0) {
    return c.json({ error: 'not found' }, 404);
  }

  const versions = [...results].sort((a, b) => versionCompare(b.version, a.version));
  return c.json({ key, versions });
});

// DELETE /items/:key/:version
itemsRouter.delete('/:key/:version', authMiddleware(2), async (c) => {
  const auth = c.get('auth');
  const key = decodeURIComponent(c.req.param('key'));
  const version = decodeURIComponent(c.req.param('version'));

  log.debug(`items/delete: world=${auth.world} key=${key} version=${version}`);

  const existing = db.prepare(
    `SELECT 1 FROM items WHERE world = ? AND key = ? AND version = ?`
  ).get(auth.world, key, version);

  if (!existing) {
    return c.json({ error: 'not found' }, 404);
  }

  const r2cfg = r2ConfigForWorld(auth.world);
  await r2DeleteObject(r2cfg, itemKeyToObjectKey(key, version));
  db.prepare(`DELETE FROM items WHERE world = ? AND key = ? AND version = ?`).run(auth.world, key, version);

  return c.json({ ok: true });
});

File: c2server/routes/items.go
package routes

import (
	"cacher/c2server/internal"
	"fmt"
	"sort"
	"strings"

	"github.com/gofiber/fiber/v2"
)

func SetupItems(app *fiber.App) {
	g := app.Group("/items")
	g.Post("/query", internal.AuthMiddleware(1), queryItems)
	g.Get("/:key", internal.AuthMiddleware(1), getItem)
	g.Delete("/:key/:version", internal.AuthMiddleware(2), deleteItem)
}

func queryItems(c *fiber.Ctx) error {
	auth := internal.GetAuth(c)

	var body struct {
		Keys []string `json:"keys"`
	}
	if err := c.BodyParser(&body); err != nil || body.Keys == nil {
		return c.JSON(fiber.Map{"items": []any{}})
	}
	if len(body.Keys) > 1000 {
		return c.Status(400).JSON(fiber.Map{"error": "max 1000 keys per request"})
	}

	var exactKeys, likePatterns []string
	for _, k := range body.Keys {
		if k == "" {
			continue
		}
		if strings.Contains(k, "*") {
			likePatterns = append(likePatterns, strings.ReplaceAll(k, "*", "%"))
		} else {
			exactKeys = append(exactKeys, k)
		}
	}

	query := "SELECT key, version, created_at FROM items WHERE world = ?"
	args := []any{auth.World}

	if len(exactKeys) > 0 || len(likePatterns) > 0 {
		var conds []string
		if len(exactKeys) > 0 {
			placeholders := strings.Repeat("?,", len(exactKeys))
			conds = append(conds, fmt.Sprintf("key IN (%s)", placeholders[:len(placeholders)-1]))
			for _, k := range exactKeys {
				args = append(args, k)
			}
		}
		for _, p := range likePatterns {
			conds = append(conds, "key LIKE ?")
			args = append(args, p)
		}
		query += " AND (" + strings.Join(conds, " OR ") + ")"
	}

	internal.LogDebug("items/query: world=%s keys=%v", auth.World, body.Keys)

	rows, err := internal.DB.Query(query, args...)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "database error"})
	}
	defer rows.Close()

	type itemRow struct {
		Key       string `json:"key"`
		Version   string `json:"version"`
		CreatedAt int64  `json:"created_at"`
	}
	byKey := map[string]itemRow{}
	for rows.Next() {
		var r itemRow
		if err := rows.Scan(&r.Key, &r.Version, &r.CreatedAt); err != nil {
			continue
		}
		if cur, ok := byKey[r.Key]; !ok || internal.VersionCompare(r.Version, cur.Version) > 0 {
			byKey[r.Key] = r
		}
	}

	items := make([]itemRow, 0, len(byKey))
	for _, r := range byKey {
		items = append(items, r)
	}
	sort.Slice(items, func(i, j int) bool { return items[i].Key < items[j].Key })

	return c.JSON(fiber.Map{"items": items})
}

func getItem(c *fiber.Ctx) error {
	auth := internal.GetAuth(c)
	key := c.Params("key")

	internal.LogDebug("items/get: world=%s key=%s", auth.World, key)

	rows, err := internal.DB.Query(
		"SELECT version, created_at FROM items WHERE world = ? AND key = ?",
		auth.World, key,
	)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "database error"})
	}
	defer rows.Close()

	var versions []internal.VersionRow
	for rows.Next() {
		var vr internal.VersionRow
		rows.Scan(&vr.Version, &vr.CreatedAt)
		versions = append(versions, vr)
	}
	if len(versions) == 0 {
		return c.Status(404).JSON(fiber.Map{"error": "not found"})
	}

	sorted := internal.SortVersionsDesc(versions)
	return c.JSON(fiber.Map{"key": key, "versions": sorted})
}

func deleteItem(c *fiber.Ctx) error {
	auth := internal.GetAuth(c)
	key := c.Params("key")
	version := c.Params("version")

	internal.LogDebug("items/delete: world=%s key=%s version=%s", auth.World, key, version)

	var exists int
	err := internal.DB.QueryRow(
		"SELECT 1 FROM items WHERE world = ? AND key = ? AND version = ?",
		auth.World, key, version,
	).Scan(&exists)
	if err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "not found"})
	}

	if err := internal.R2DeleteObject(auth.World, internal.ItemKeyToObjectKey(key, version)); err != nil {
		internal.LogError("items/delete: r2 delete failed: %v", err)
	}
	internal.DB.Exec("DELETE FROM items WHERE world = ? AND key = ? AND version = ?", auth.World, key, version)

	return c.JSON(fiber.Map{"ok": true})
}
